### ISSUE TITLE
Finish SPIR-V emitter for MojoShader

### ISSUE THREAD

**Issue by flibitijibibo** [flibitijibibo/flibitBounties#1]
**Introductory Information:**

[MojoShader](https://icculus.org/mojoshader/) is a project that takes in Direct3D shader bytecode and emits shaders for other graphics APIs, including ARB1, GLSL, and Metal shaders. It is used by FNA for its Effects framework implementation and is used by every single game using [FNA](http://fna-xna.github.io/), as well as other non-XNA games like Super Meat Boy, Psychonauts, and Killing Floor.

[SPIR-V](https://www.khronos.org/registry/spir-v/) is the Khronos Group's standard IR for graphics/compute shaders, used by Vulkan, OpenCL, and OpenGL (via [ARB_gl_spirv](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_gl_spirv.txt)).

**The Project:**

We need to implement a SPIR-V emitter for MojoShader which works with both OpenGL as well as Vulkan. Currently our priority is OpenGL support since we have all the infrastructure (mostly) set up to work with `ARB_gl_spirv` program objects already. In the long term, this is primarily for [FNA's VulkanDevice](https://github.com/FNA-XNA/FNA/issues/11), though it is applicable to any game currently using MojoShader that properly uses MOJOSHADER_glBestProfile to autodetect usable shader profiles.

Currently, the MojoShader repo FNA works on is [here](https://github.com/FNA-XNA/MojoShader). However, the current active repository for the SPIR-V work is [here](https://bitbucket.org/krolli/mojoshader/src/spirv/). The part you care about is in `profiles/mojoshader_profile_spirv.c`.

For the most part the SPIR-V emitter works, but we are missing many opcodes (check the file for `EMIT_SPIRV_OPCODE_UNIMPLEMENTED_FUNC`) and we need to be at least up to feature parity with the glsl120 emitter. glsl120 as well as the other emitters in the profiles folder will be helpful as references. A checklist can be found [below](https://github.com/flibitijibibo/flibitBounties/issues/1#issuecomment-494190160).

Lastly, the whole file needs at least one final pass, for polish. Style cleanup would be nice but isn't mandatory. 

**Prerequisites:**

MojoShader is written in C99, so you'll need to be reasonably comfortable with that language to finish this project. Experience with graphics programming is also a plus, but you won't be doing _too_ much of that since you're just touching shader content and not stuff like GL callstreams and what have you.

Knowledge of the [Direct3D Shader Model 3.0 specification](https://msdn.microsoft.com/en-us/library/windows/desktop/bb509656(v=vs.85).aspx) is highly recommended unless you can read through a whole lot of documents on assembly-like languages very thoroughly. You'll mostly want to read up on the register sets ([vs](https://msdn.microsoft.com/en-us/library/windows/desktop/bb172963(v=vs.85).aspx), [ps](https://msdn.microsoft.com/en-us/library/windows/desktop/bb172920(v=vs.85).aspx)) and instruction sets ([vs](https://msdn.microsoft.com/en-us/library/windows/desktop/bb172938(v=vs.85).aspx), [ps](https://msdn.microsoft.com/en-us/library/windows/desktop/bb219854(v=vs.85).aspx)). MojoShader already handles the bytecode parsing but you'll want to know (at a high level, at least) what the emitter functions are referencing.

You'll also want to read up on the [SPIR-V specification](https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.pdf), for pretty much the same reasons as the Direct3D shader spec.

**Example Games:**

At a basic level, lots of FNA games use [these Effect files](https://github.com/FNA-XNA/FNA/tree/master/src/Graphics/Effect/StockEffects) which should work with MojoShader's [testparse](https://github.com/FNA-XNA/MojoShader/blob/fna/utils/testparse.c) program, so if you want a quick set of test shaders, this is probably a good place to start. Just about every XNA game uses SpriteEffect; it'd probably be easier to write a list of games that _don't_ use it if I could even remember any.

Examples of more complex Effects include those found in FEZ and Reus, with moderately complex shaders found in Skulls of the Shogun, Apotheon, Cryptark, Escape Goat 2, Murder Miners, and Gateways.

With some LD_PRELOAD magic it may be possible to test SPIR-V support with the Linux/macOS versions of Super Meat Boy and Psychonauts, but this has been untested - maybe ask Ryan C. Gordon about this if you want to hack on those games.

**How Much Can flibit Help?**

I might be able to help with `ARB_gl_spirv` integration and maybe help clear up questions about the D3D spec, but I'll confess to not having the best knowledge of the SPIR-V spec. By the time you get test shaders working you'll probably already know more about SPIR-V than me.

**Budget/Timeline:**

Measuring in weekends, I expect this to take about **2 weekends**. The first weekend would be spent on scribbling out the remaining opcodes and the second weekend would be spent on testing/polish. I currently have **$500 USD** allocated for this project.

**Comment by angusholder** [comment]
I've been going through MojoShader and API documentation and think I can take this on. I've been able to build the project and run through those example shaders you mentioned from FNA, have gotten testparse.c working and added SPIRV disassembly to it, and now am working on argument swizzles to get ADD working for a start

**Comment by flibitijibibo** [comment]
Doing a weekly update for this even though it's only been 5-6 days - going to try and keep all these on Friday.

I eventually (sorry!) replied to the DMs, feel free to ping me if something else comes up. I can also get in touch with meklu if we need to figure out what the existing source was doing.

**Comment by angusholder** [comment]
It's going well now, thanks for the help!

I'm working on it now and have just got ADD, DP3, and DP4 working, along with some plumbing they rely on like uniforms, temp regs stored as globals, and some of the source and dest modifiers. Now that I've worked out the plumbing I needed, putting in the most of the other arithmetic instructions ought to be a breeze. I'm sure I'll encounter some head-scratchers when I get onto samplers though!

**Comment by flibitijibibo** [comment]
Weekly checkup as usual - we've been in contact throughout the week, just need an overall "what done happened since last Friday" to keep track of what's going on. This next week might be a good time to look in Ryan since he's ultimately the one merging this in.

**Comment by angusholder** [comment]
So I've implemented the majority of the arithmetic instructions, have linked up (hopefully) all the builtin attributes (the equivalents of `gl_PointSize`, `gl_Position`, etc..) and passing attribs from vertex shader to pixel shader, I now handle creating samplers, and have implemented a couple of texture sampling instructions (TEXLD & TEXLDB).

**Comment by flibitijibibo** [comment]
Doing the weekly update, will save Angus some time this week by just pointing to the diff tracker...

https://github.com/flibitijibibo/MojoShader/pull/6

It's now at the point where basic shaders _should_ work but have only been put through the validator and not a real driver. If you have some shaders, an ARB_gl_spirv driver, and a dash of bravery, feel free to give this branch a try and post what you find.

**Comment by flibitijibibo** [comment]
Forgot to do weekly updates! Switch work had me busy all weekend, but now I can get back into the generic multiplat stuff. Anything new with this?

**Comment by angusholder** [comment]
Hey Ethan, sorry I haven't had so much time recently, I managed to get everything hooked up compile-time-wise for using my Mojoshader branch with FNA, at runtime it fails to compile the bytecode though, so my next step is to see what SPIRV glslangValidator produces for equivalent GLSL to see if I'm missing something there, as currently SPIRV-val gives no complaints for what I'm generating for SpriteEffect. I should think I'll be able to get back onto it within the next few days.

**Comment by flibitijibibo** [comment]
Digging this up... any luck with this lately? Thankfully this is the one bounty that isn't urgent, so we're not pressed for time, but wanted to make sure there weren't any walls that I could help look into.

**Comment by flibitijibibo** [comment]
Opening this back up - let me know if you can get back into this at some point.

**Comment by vitorhnn** [comment]
I was messing around with this and I managed to get SpriteEffect's pixel shader working using a test program (not FNA) by using testparse to dump the raw SPIR-V and using ARB_gl_spirv and a vertex shader with compatible location specifiers to load it, which was unexpected, because the disassembly doesn't contain any Location decorators for the color output and glslangValidator requires those to even compile. Maybe this is just Nvidia's driver being lenient, I don't know.

Uniforms also _should_ require Location decorators, which leads me to believe the storage class being used should be changed from Uniform to UniformConstant, as the spec doesn't allow applying Location decorators to the Uniform storage class.

Having shader-defined locations for uniforms and samplers allows us to completely toss away the glGetUniformLocation and glGetAttribLocation reflection currently being done on mojoshader_opengl.c as we can just map DXBC's register numbers to a fixed location, but samplers might complicate that a bit.

**Comment by flibitijibibo** [comment]
Wouldn't surprise me if NVIDIA's SPIR-V compiler was as lenient as the GLSL, though I wonder if they have warnings somewhere. Maybe the Mesa SPIR-V compiler says something interesting?

Luckily the reflection is pretty easy to skip if we know the `ctx` is SPIR-V, we can write `impl_SPIRV_GetSamplerLocation` or what have you to bypass all that.

**Comment by infapi00** [comment]
> Wouldn't surprise me if NVIDIA's SPIR-V compiler was as lenient as the GLSL, though I wonder if they have warnings somewhere. Maybe the Mesa SPIR-V compiler says something interesting?

Hi, I'm Alejandro Piñeiro, one of the developers working on ARB_gl_spirv support for Mesa. Unfourtunately, this is still a WIP. Although we have been pushing to master upstream several patches during the past weeks, there are still work to do. Specifically, UBO/SSBO support needs some finishing and cleaning.

In any case, if you are interested on testing it with Mesa, the last version of the development branch is the following one:
https://github.com/Igalia/piglit/tree/igalia/arb_gl_spirv

That one gets the following stats then executing the shaders from the piglit test suite on SPIR-V
    [34453/34453] skip: 5877, pass: 28353, fail: 210, crash: 13

So although it is still wip, it should cover most of the functionality required (and FWIW, the proprietary NVIDIA drivers gets worse numbers).

FWIW, I subscribed to this issue because eventually we would like to test our driver against a real application, instead of relying on specification tests. So I would appreciate any comment when the progress on this issue is enough to do some testing, Thanks.

**Comment by infapi00** [comment]
> In any case, if you are interested on testing it with Mesa, the last version of the development branch is the following one:
> https://github.com/Igalia/piglit/tree/igalia/arb_gl_spirv
Sorry wrong link, this is the correct one:
https://github.com/Igalia/mesa/tree/igalia/arb_gl_spirv

**Comment by flibitijibibo** [comment]
Mesa is definitely a high-priority target for us as it'll let us target core profile without having to update the GLSL emitters to support specific versions (all we've got is 1.20). I do see gen7+ but just to be sure, is this ready for Haswell too? I do still have my Dev Days kit so I can try this out locally on real hardware as it develops.

**Comment by infapi00** [comment]
>  I do see gen7+ but just to be sure, is this ready for Haswell too?

Right now our priority is get it working on gen8+. Having said so, my main development laptop is haswell, and we didn't find any gen7 specific issue yet. So it is really likely that getting it working on gen8+ would get it working (or mostly working) on gen7 too.

**Comment by vitorhnn** [comment]
I've managed to get [SpriteEffect](https://ptpb.pw/fmVq.png) working, though this is a test program that doesn't use MojoShader's Effect framework, since that requires more work on mojoshader_opengl.c to work, particularly uniforms (some tips here would be quite helpful).

I'd also like to apologize for the lack of updates, my midyear break is over and The International is ongoing, so I'm slacking off horribly.

**Comment by flibitijibibo** [comment]
What does the sample's call stream look like compared to mojoshader_opengl.c's? I was thinking all the work would just be in setup and the existing call stream would be fine, but if we know what the difference is between the two it should help us figure out what changes need to be made. Luckily the Effect stuff won't be affected, just the GL calls.

**Comment by krolli** [comment]
Hi, I've been playing around with code I found in [vitorhnn's repo](https://github.com/vitorhnn/MojoShader). Right now I wanted to get all the FNA shaders to generate at least some spir-v without errors to get a grasp on how the spir-v emitter is implemented. I'm almost done with that so next step would be to start getting it working within actual games (or maybe test it with some test programs before that, not yet decided).

I've been wondering, since there have been multiple people working on this, the fork-chain is getting a bit long and I'm not sure whether I should create another repo for my changes or not.

**Comment by flibitijibibo** [comment]
I wouldn't worry too much about the fork chain - at the end the pull request is going straight to mainline, the chain just lets us know what the whole history is.

**Comment by krolli** [comment]
Great, so as long as there are no forks I'm not aware of, i'll go ahead and continue on from there.

**Comment by flibitijibibo** [comment]
krolli's working on this slowly but surely! Marking as In Progress.

**Comment by krolli** [comment]
Since there has been some reasonable breakthrough, I thought I'd post some update.

I can currently convert all of the .fxb files in FNA without any SPIR-V validation complaints. I have SpriteEffect working in the simplest application (just drawing some sprites) and I have moved on to getting an actual game running on this. OpenGL integration is now in place as well. Uniform management is same as in GLSL back-end (uniforms packed into large arrays) but I have noticed some uniform location issues in shaders. For now, I managed to work around them, but I expect more complicated shaders to run into this again, requiring real fix.

Game I'm testing right now is Bastion since I already have it and I figured it would be fairly simple in terms of shaders being used. Menus are working fine, but once I enter actual gameplay, it gets a bit worse. Things seems to be rendered at much lower resolution and frame time also goes up considerably compared to GLSL (GLSL 16.6ms likely vsync limited, SPIR-V 59ms). From frame debugging GLSL version, I suspect low resolution problem to be somehow tied to bloom or one of the other post-processes.

![bastion-spirv](https://user-images.githubusercontent.com/6864891/49699187-7932b080-fbce-11e8-8514-66c5e71fe43b.png)
![bastion-spirv-bad](https://user-images.githubusercontent.com/6864891/49699188-7c2da100-fbce-11e8-87fe-d9ab94a12f49.png)

Unfortunately, attempting to debug or profile SPIR-V version is a bit of a problem right now. Most recent changes that got Bastion working have also caused RenderDoc to crash whenever I attempt to open a capture. Based on minidump, it seems to be crash in whatever is being used to disassemble SPIR-V. I also tried NSIGHT, but it refused to do anything due to calls to glSpecializeShaderARB (required to use SPIR-V modules). I'll probably just grab RenderDoc source code and see if I can fix that crash or, if that doesn't work, work around it in mojoshader SPIR-V generator.

**Comment by flibitijibibo** [comment]
The BloomComponent (used in a bunch of games including Bastion) does half-size rendering with a GaussianBlur effect, so I wonder if the issue is the sampler parameters - maybe it’s using nearest-neighbor for some reason?

For RenderDoc, if it’s not already set: FNA_OPENGL_FORCE_CORE_PROFILE=1 might fix any capture issues.

**Comment by krolli** [comment]
> The BloomComponent (used in a bunch of games including Bastion) does half-size rendering with a GaussianBlur effect, so I wonder if the issue is the sampler parameters - maybe it’s using nearest-neighbor for some reason?

I don't think sampler parameters are at fault here. I didn't have to do anything different from GLSL for samplers to work in other shaders (though maybe there are some differences I didn't notice). Also, there doesn't seem to be any blur beyond that introduced when you stretch smaller image on larger one. I am more inclined to think that blur passes are broken in some way, leaving image sharp, and that when blurred version is drawn on top of original, alpha gets set to 1 and overwrites the full res version. But I'm only guessing and I'd rather have working debugger, especially since there might be pretty rough problems once I get to 3D effects.

> For RenderDoc, if it’s not already set: FNA_OPENGL_FORCE_CORE_PROFILE=1 might fix any capture issues.

Thanks, didn't know about that. It fixed the warning about unsupported compatibility profile, but sadly didn't help with crash when opening the capture. It worked before my last batch of changes that introduced vpFlip uniform and fixed some missing attribute decorations so I suspect something there is causing problems. However, spirv-dis has no trouble disassembling any of my generated shaders, so if RenderDoc is using some older version of SPIRV-Tools for its disassembler, linking against newer version might be enough to fix the problem.

**Comment by flibitijibibo** [comment]
Not sure if it's relevant but I was looking over some Proton stuff and ran into some issues filed in the DXVK repo related to NVIDIA's newest compiler and SPIR-V performance taking a nosedive - does setting `__GL_NextGenCompiler=0` improve performance at all?

**Comment by krolli** [comment]
> does setting __GL_NextGenCompiler=0 improve performance at all?

Doesn't seem to have any effect. Tried with both 1 and 0 and perf. seems to be the same.

I also looked into RenderDoc sources. Unfortunately, crashing code is RenderDoc stuff (not 3rd party lib) and it doesn't seem to be a simple fix. It looks like it doesn't support shaders using UniformConstant and assumes everything is either Uniform Block or Sampler and adding support for that will probably require some integration into UI as well. Maybe @baldurk would be able to shed some more light on this. It's possible I misunderstood things.

I haven't yet looked at what would be needed to use Uniform instead of UniformConstant in MojoShader, but it seems like a possible way to solve this. It wouldn't surprise me if drivers weren't used to seeing UniformConstant either, so they might be generating worse code for it.

**Comment by baldurk** [comment]
Yes the SPIR-V support is not using SPIRV-Tools at all, and it was written only for Vulkan so it assumes reasonably modern shaders (constant buffers, etc). There may be gaps for GL-only constructs like standalone uniforms which I can fix with a repro case, hopefully it shouldn't be too bad.

When I added SPIR-V support for GL it was a blind implementation, I didn't know of any non-toy applications using it, so the test I set up also used non-legacy shaders.

**Comment by krolli** [comment]
I could provide a bunch of SPIRV binaries (with GLSL and D3D counterparts) if you'd be interested in expanding support for GL+SPIRV. MojoShader SPIRV backend is still a work in progress and there might be some bugs in terms of functionality, but I try to keep all of them in shape that at least passes validation.

**Comment by flibitijibibo** [comment]
As a starting point, here are the built-in effects for FNA:

https://github.com/FNA-XNA/FNA/tree/master/src/Graphics/Effect/StockEffects
https://github.com/FNA-XNA/FNA/tree/master/src/Graphics/Effect/YUVToRGBA

The SPIR-V branch should be able to parse SpriteEffect at least, haven't checked out the others yet.

**Comment by krolli** [comment]
It should be able to go through all of them. Here is a zip with generated shaders for different back-ends (makes it easier to see what is going on and helps with spotting bugs)
[data.zip](https://github.com/flibitijibibo/flibitBounties/files/2673674/data.zip)

fxb - compiled effect files (also has raw SPIRV modules mixed in)
d3d - disassembled effect files
glsl - output of MojoShader GLSL back-end
spirv - output of MojoShader SPIRV back-end
spvasm - disassembled raw SPIRV modules from fxb folder

**Comment by krolli** [comment]
I think I've found out where the problem was. It seems obvious in hindsight, but when assigning locations to uniforms, I only took into account current shader and didn't account for locations being shared between shader stages as well.

![bastion-spirv-good-0](https://user-images.githubusercontent.com/6864891/50055565-05098700-0151-11e9-9021-15d385a8b82b.png)
![bastion-spirv-good-1](https://user-images.githubusercontent.com/6864891/50055566-05098700-0151-11e9-9ad2-ee152594f98f.png)

Now I just need to rewrite the thing so locations are actually assigned when program is linked. Good news is, this seems to have been the cause of both visual and perf problems.

**Comment by baldurk** [comment]
I [pushed a fix](https://github.com/baldurk/renderdoc/commit/5f96644d08b4d5cad04e10c9a22ff405dfdb2948) which should let these shaders work. I fixed a couple of GL-isms: textures without explicit bindings, and bare uniforms not in buffers.

If you run into any more cases, drop me the SPIR-V and I should be able to fix it.

**Comment by krolli** [comment]
Thanks, @baldurk. Just tested it and it seems to be working pretty well. Will definitely let you know if I run into more issues.

**Comment by flibitijibibo** [comment]
Checking the pulse on this one... at this point it's down to getting all the instructions, right? We should probably get a checklist written up so we know what's needed and what's already done.

**Comment by krolli** [comment]
There is one final hurdle before it will be down to that, which is a bit tricky and I haven't had much time to dive into it.

Uniform locations in spir-v need to be unique in final linked program and have to take into account size of arrays and such. Basically I need to rewrite location assignment to be postponed until linking for either pixel or vertex shaders. I haven't seen this in any of the other mojoshader back-ends but I think I know how to do it. Sadly, I've been rather busy with other things lately.

After that, I don't know of any other issue apart from missing instructions, but we'll see what we run into when throwing more complicated games at it.

**Comment by krolli** [comment]
Update: Finally found time to finish the link-time location assignment. Everything still seems to be working fine.
I will see what other games using FNA or mojoshader I have and maybe test those to see if I run into some other issues that might complicate things. Beyond that, there seem to be around 40 missing op-codes left to implement: BEM, BREAK, BREAKC, BREAKP, CALL, CALLNZ, CND, DST, DSX, DSY, ELSE, ENDIF, ENDLOOP, ENDREP, EXPP, IF, IFC, LABEL, LIT, LOGP, LOOP, REP, RESERVED, RET, SETP, TEXBEM, TEXBEML, TEXCRD, TEXDEPTH, TEXDP3, TEXDP3TEX, TEXLDD, TEXLDL, TEXM3X2DEPTH, TEXM3X2PAD, TEXM3X2TEX, TEXM3X3, TEXM3X3PAD, TEXM3X3SPEC, TEXM3X3TEX, TEXM3X3VSPEC, TEXREG2AR, TEXREG2GB, TEXREG2RGB.

**Comment by flibitijibibo** [comment]
I believe a lot of those are tested by this...

https://github.com/FNA-XNA/MojoShader/tree/fna/tests

... but I had to comment this block to get the test to work :|

https://github.com/FNA-XNA/MojoShader/blob/fna/mojoshader.c#L9070

And some of them can be pretty lazy, like EXPP/LOGP:

https://github.com/FNA-XNA/MojoShader/blob/fna/mojoshader.c#L3810

Also, good news! If you're absolutely sick of working on a giant file, we're in the process of fixing that and migrating the SPIR-V stuff to this layout should be nice and simple:

https://github.com/FNA-XNA/MojoShader/pull/5

**Comment by krolli** [comment]
Size of the file is not that much of a problem (I have survived bigger beasts :P ), though I am thinking of ditching or redoing some macros that generate function signatures. They tend to mess up jump-to-definition in simpler editors and I'm usually not a fan of hiding code in macros this way.

**Comment by flibitijibibo** [comment]
We’re fighting the same issue - the only thing I’m worried about is accidentally bloating the binary size (which our current patch totally does). I imagine performance issues aren’t going to show up so size will be our focus for optimization.

**Comment by TheSpydog** [comment]
Hey @krolli I've been looking over your branch and it's really impressive how much you've gotten done! A couple questions/suggestions:

1) Is there anything that could be done to reduce (or remove) the spirv struct inside Context? In the separate-profile-files PR branch above, the context struct is defined in a header and shared with all profile emitters, so I'd prefer to keep it as lean as possible. Maybe we could split it out into a static, standalone `spvContext` struct or something that could just live inside a `mojoshader_profile_spirv.c` file?

2)  I've noticed there are a lot more `spv_*` helper functions than for any other emitter. I wonder if we could replace some of them with more generic helper methods, or maybe even remove the need for some of them? Especially the simple ones like `spv_emit_word`, since essentially all it does is call `buffer_append`.

Anyways, keep up the good work! 🙂

**Comment by krolli** [comment]
Hi @TheSpydog ,
thanks, but a lot of the work was done by people before me. As for your questions:

1. There are a few things that can be done about it. The definition of it certainly doesn't need to be inside Context. I believe noone expected it to get that big initially. All of the usual ways of decoupling those two are doable, but as I was implementing branches and loops, I realized I sometimes used existing members of Context and sometimes needed to expand them with extra information (branch_labels_stack and branch_labels_patch_stack are examples of that). I am wondering whether it might make sense to have a union / opaque pointer to emitter data in Context, but I haven't looked into other emitters to see how useful it would be for them.
2. It is definitely possible and I have some idea how to rewrite type id tracking to make it require less code and be easier to work with. I want to get my current testing set working without any asserts/validation failures firing first though. `spv_emit_word` along with `spv_emit`, `spv_emit_part` and `spv_emit_part_va` will likely stay though. The most likely candidates for pruning are `spv_get{type}` functions, just need to redo the whole type id tracking.

I assume this is related to your work on splitting emitters into separate files. I was actually waiting for that change to land in master and then do the separation on spirv branch myself, but I noticed the change is in different repo. Not sure who will be merging between those two.

**Comment by flibitijibibo** [comment]
The big refactor is in:

https://hg.icculus.org/icculus/mojoshader/
https://github.com/FNA-XNA/MojoShader

Hopefully the isolation of the SPIR-V emitter will make this a little less scary to look at since it's not part of a 10kLOC file >_>

**Comment by krolli** [comment]
Comparing https://github.com/flibitijibibo/MojoShader and https://github.com/FNA-XNA/MojoShader, I am starting to think I should switch spirv stuff over to one of the other repos. Personally, I am more used to mercurial than git, but I will switch to whichever is more convenient for you to deal with.

**Comment by flibitijibibo** [comment]
Mercurial would actually be better, since we're ultimately going to push this to upstream before pulling it into FNA's repository.

**Comment by krolli** [comment]
I moved the changes to mercurial repo. You can find it over at bitbucket (https://bitbucket.org/krolli/mojoshader). I did my best to preserve all previous commits, though some changes (CMakeLists.txt files that were vastly different) have been lost. As a result, there are a few commits that look empty as they only made changes to those files.

Now, on to making it work again in new repo and then merging all the changes from master in last year or so. :)

**Comment by TheSpydog** [comment]
So what are the next steps with this? Just filling in the missing opcodes and testing? I’ve noticed that some of the other emitters don’t emit anything for obscure opcodes so maybe it’d be good to get a list of the ones we actually need to finish this up. I’ll have some time after this week to help finish this up if you’d like.

**Comment by TheSpydog** [comment]
I looked at the current source, and it appears we still need to fill in ~26~ ~13~ ~9~ ~8~ ~7~ ~5~ 4 opcodes to achieve parity with the GLSL emitter:

*LAST UPDATED: August 28, 2019, 14:00 EDT*

- [ ] [TEXBEML](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/texbeml---ps)
- [ ] [TEXM3X3TEX](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/texm3x3tex---ps)
- [ ] [TEXM3X3SPEC](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/texm3x3spec---ps)
- [ ] [TEXM3X3](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/texm3x3---ps)

Beyond this list, there are 9 other unimplemented opcodes, but since the GLSL emitter doesn't implement them either... 🤷‍♂

**Comment by flibitijibibo** [comment]
After looking over the changelog it only occurred to me (read: @TheSpydog reminded me) how many people have worked on this over time, and since we're looking to be close to the "end" (still need to review the working changes but at this point we're happy to just have it work, cleanup can happen during review), I'd like to start distributing the bounty (which now has a larger budget btw) so we can simplify the remaining work for anyone that wants to hammer out the rest of the busy work. @meklu, @angusholder, and @krolli, send a quick e-mail to the address at the bottom of [this page](http://www.flibitijibibo.com/index.php?page=Home) so I can start sending these payments out. I can then amend the OP as needed.

**Comment by krolli** [comment]
Hi, sorry for silence last couple of weeks.

@TheSpydog I've been implementing instructions only when they appear in shaders I test. In other words, I didn't implement anything I can't immediately run through validator. I have a separate repo (https://bitbucket.org/krolli/mojoshader-spirv/src/default/) where I keep stuff for testing. If you'd like to use that, let me know.

Anyway, I'm currently trying to get https://github.com/FNA-XNA/MojoShader/blob/fna/tests/1.vsh working, for which I currently get following errors (assuming check for uninitialized registers is disabled as mentioned before):
```
ERROR: BREAKP unimplemented in spirv profile
ERROR: CALL unimplemented in spirv profile
ERROR: CALLNZ unimplemented in spirv profile
ERROR: LABEL unimplemented in spirv profile
ERROR: RET unimplemented in spirv profile
ERROR: SETP unimplemented in spirv profile
ERROR: register type p0 is unimplemented
ERROR: unsupported source modifier 13
```
I plan to have a look at those errors next, which will probably involve doing some infrastructure for function calls.

I suspect that after opcodes, we will still run into some issues with various missing things, but hopefully nothing complicated.

**Comment by flibitijibibo** [comment]
I just finished rewriting the bounty OP - for those familiar with this project already it's nothing really new, it just tweaks things to make it clear that we're now in the latter half of the project.

**Comment by flibitijibibo** [comment]
Something interesting for VPOS, if it hasn't been done already:

https://github.com/FNA-XNA/MojoShader/issues/14

The Vulkan/SPIR-V specification is thankfully very explicit about FragCoord:

https://vulkan.lunarg.com/doc/view/1.0.26.0/linux/vkspec.chunked/ch14s06.html

D9VK simply loads the value and subtracts the pixel center: https://github.com/Joshua-Ashton/d9vk/blob/bc64e8e0263004908090da60b4c478a6573fe7d0/src/dxso/dxso_compiler.cpp#L1111

**Comment by flibitijibibo** [comment]
Since I'm here, updated the checklist and included links to every instruction on MSDN.

**Comment by flibitijibibo** [comment]
A quick note for TEXLDD... it seems like even GLSL never ran into this. The emitter for this instruction is just some added setup for extensions like `EXT_gpu_shader4` followed by a call to `glsl_texld`, which hard asserts on the ldd path:

https://github.com/FNA-XNA/MojoShader/blob/fna/profiles/mojoshader_profile_glsl.c#L1665

So we probably don't have any data for this, but on the other hand a lot of why the feature wasn't implemented was because of GLSL itself. I think the SPIR-V path is essentially the same as any image sample, but with the Grad (0x4) flag set and src2/src3 pulled in when `texldd == 1` in `spv_texld`.

After that's done (or at least documented?), unless someone really wants to do the remaining SM1 instructions, I think we're at the point where we can start cleaning up and make a beta build for people to try out and review.

**Comment by krolli** [comment]
I reached the point where all the shaders I have at hand run through testparse without issues, including UT3 stuff and ignoring CTAB issues on Psychonauts. There are still some opcodes and source modifiers which are present in GLSL that I never hit so far. If someone is sitting on some large stack of shader binaries, now would be good time to throw them in my direction.

I know @TheSpydog had some issue with Chasm on Intel GPUs, but renderdoc capture he sent me looked fine on my end, when viewed on nvidia GPU. I suspect this might be a driver issue, but hard to say without seeing where it went wrong on his machine.

I think we can start working on some testing build. I could try to implement remaining pieces blindly, without running output through validator, but I'd rather have it barf an assert/error on unimplemented code path, than cause a driver crash on badly implemented one. I can also try to create test cases for some of those remaining code paths, but it might be easier to find examples of it in real shipped games.

**Comment by flibitijibibo** [comment]
Sounds good, I’ll do a cleanup pass on Monday, aiming for a public test on Wednesday.

**Comment by flibitijibibo** [comment]
@krolli Whenever you're free, can you do one more rebase against default? I'll try to send a big cleanup patch to go on top of that, and after everything's merged in your repo I'll put together some test games and post this to my plan file.

My current testing list:

- FEZ
- Dust: AET
- Murder Miners
- Escape Goat 2
- Capsized
- Apotheon
- Cryptark

I would probably put Bastion and Reus in there (Reus especially) but those are both internally maintained now - I'll have to see if simply swapping out FNA binaries still works for those. EDIT: Same for Celeste, good catch Caleb!

**Comment by TheSpydog** [comment]
I recommend adding Celeste to that list. Did some testing a few days ago and noticed it still had some major issues on Intel when dashing. Not sure if it affects other vendors as well but might be worth a look.

**Comment by krolli** [comment]
Alright, shall I also collapse all changesets while rebasing or should I try to keep as much history as possible? I will still keep it on spirv branch, if you don't mind, just so we can easily switch between the two.

Also, I just noticed "Replace glProgramViewportFlip with glProgramViewportInfo" change on default. It seems to have affected Chasm CRT shader. Will probably need to mirror that in spirv profile as well.

**Comment by flibitijibibo** [comment]
The final commit to upstream will be a single commit (with a whole boatload of credits) so you can squash it if it makes life easier.

I think Chasm’s CRT shader uses the broken VPOS, I’ll go ask James... EDIT: Yup it's VPOS... we're actually accurate to XNA now and it makes the shader less pretty, go figure.

**Comment by krolli** [comment]
Alright, things have been rebased and collapsed. Its in spirv-rebase branch and I closed spirv branch. I might go over changes done in other files and try to reduce some noise (eg. whitespace changes) later, unless patch you've mentioned already does that.

**Comment by flibitijibibo** [comment]
Here's the first of probably two cleanup patches. This cleans up every file except mojoshader_profile_spirv.c. It mostly fixes style, but does cut out a bunch of debug stuff to reduce clutter (since this patchset is going to be huge no matter how we div it up).

Notes aside from cleanup:
- Probably have to implement the vpos flip variable (no need to adjust half-pixel!)
- Can we remove the ARB program paths from SPIR-V? Seems unlikely that a driver will support ARB_gl_spirv but not GL2...

Expect the second patch later today.

```
# HG changeset patch
# User Ethan Lee <flibitijibibo@flibitijibibo.com>
# Date 1565792958 14400
#      Wed Aug 14 10:29:18 2019 -0400
# Branch spirv-rebased
# Node ID 1041c1e582135efa39a623bae87c238f287c6ef1
# Parent  73770d3b26ae1632e9778cd61ac5f8b0fbf889d0
spirv cleanup part 1

diff -r 73770d3b26ae -r 1041c1e58213 mojoshader.c
--- a/mojoshader.c	Wed Aug 14 08:07:15 2019 +0200
+++ b/mojoshader.c	Wed Aug 14 10:29:18 2019 -0400
@@ -3462,8 +3462,8 @@
         if (strcmp(retval->profile, "spirv") == 0)
         {
             int binary_size = retval->output_len - sizeof(SpirvPatchTable);
-            uint32* binary = (uint32*)retval->output;
-            SpirvPatchTable* table = (SpirvPatchTable*)&retval->output[binary_size];
+            uint32 *binary = (uint32 *) retval->output;
+            SpirvPatchTable *table = (SpirvPatchTable *) &retval->output[binary_size];
 
             if (table->vpflip.offset)      binary[table->vpflip.offset]      = table->vpflip.location;
             if (table->array_vec4.offset)  binary[table->array_vec4.offset]  = table->array_vec4.location;
@@ -3475,8 +3475,8 @@
                 SpirvPatchEntry entry = table->samplers[i];
                 if (entry.offset)
                     binary[entry.offset] = entry.location;
-            }
-        }
+            } // for
+        } // if
 #endif
 
         // we don't own these now, retval does.
diff -r 73770d3b26ae -r 1041c1e58213 mojoshader_common.c
--- a/mojoshader_common.c	Wed Aug 14 08:07:15 2019 +0200
+++ b/mojoshader_common.c	Wed Aug 14 10:29:18 2019 -0400
@@ -1062,7 +1062,7 @@
         } // while
     } // if
 
-    const uint8 *data = (const uint8 *)_data;
+    const uint8 *data = (const uint8 *) _data;
     size_t write_pos = start - pos;
     size_t write_remain = len;
     size_t written = 0;
@@ -1079,7 +1079,7 @@
         write_pos     = 0;
         item          = item->next;
     } // while
-}
+} // buffer_patch
 
 // Based on SDL_string.c's SDL_PrintFloat function
 size_t MOJOSHADER_printFloat(char *text, size_t maxlen, float arg)
diff -r 73770d3b26ae -r 1041c1e58213 mojoshader_effects.c
--- a/mojoshader_effects.c	Wed Aug 14 08:07:15 2019 +0200
+++ b/mojoshader_effects.c	Wed Aug 14 10:29:18 2019 -0400
@@ -849,24 +849,6 @@
     const uint8 *ptr = (const uint8 *) buf;
     uint32 len = (uint32) _len;
 
-    #if 0
-    {
-        unsigned long long checksum = 0x2394F7C94D7A6E13ull;
-        for (uint32 i = 0; i < len; i++)
-        {
-            checksum = (checksum ^ ptr[i]) * 13;
-        }
-        char pathbuf[128];
-        snprintf(pathbuf, sizeof(pathbuf), "mojoshader-binaries/%llX.fxb", checksum);
-        FILE *fp = fopen(pathbuf, "wb");
-        if (fp)
-        {
-            fwrite(ptr, 1, len, fp);
-            fclose(fp);
-        }
-    }
-    #endif
-
     /* Supply both m and f, or neither */
     if ( ((m == NULL) && (f != NULL)) || ((m != NULL) && (f == NULL)) )
         return &MOJOSHADER_out_of_mem_effect;
diff -r 73770d3b26ae -r 1041c1e58213 mojoshader_opengl.c
--- a/mojoshader_opengl.c	Wed Aug 14 08:07:15 2019 +0200
+++ b/mojoshader_opengl.c	Wed Aug 14 10:29:18 2019 -0400
@@ -433,11 +433,11 @@
 } // impl_GLSL_MaxUniforms
 
 #if SUPPORT_PROFILE_SPIRV
-static const SpirvPatchTable* spv_getPatchTable(MOJOSHADER_glShader* shader)
+static const SpirvPatchTable* spv_getPatchTable(MOJOSHADER_glShader *shader)
 {
-    const MOJOSHADER_parseData* pd = shader->parseData;
+    const MOJOSHADER_parseData *pd = shader->parseData;
     size_t table_offset = pd->output_len - sizeof(SpirvPatchTable);
-    return (const SpirvPatchTable*)(pd->output + table_offset);
+    return (const SpirvPatchTable *) (pd->output + table_offset);
 } // spv_getPatchTable
 
 static int spv_CompileShader(const MOJOSHADER_parseData *pd, int32 base_location, GLuint *s)
@@ -446,12 +446,12 @@
 
     GLsizei data_len = pd->output_len - sizeof(SpirvPatchTable);
     const GLvoid* data = pd->output;
-    uint32* patched_data = NULL;
+    uint32 *patched_data = NULL;
     if (base_location)
     {
-        patched_data = (uint32*)Malloc(data_len);
+        patched_data = (uint32 *) Malloc(data_len);
         memcpy(patched_data, data, data_len);
-        const SpirvPatchTable* table = (const SpirvPatchTable*)&pd->output[data_len];
+        const SpirvPatchTable *table = (const SpirvPatchTable *) &pd->output[data_len];
         if (table->vpflip.offset)      patched_data[table->vpflip.offset]      += base_location;
         if (table->array_vec4.offset)  patched_data[table->array_vec4.offset]  += base_location;
         if (table->array_ivec4.offset) patched_data[table->array_ivec4.offset] += base_location;
@@ -462,10 +462,10 @@
             SpirvPatchEntry entry = table->samplers[i];
             if (entry.offset)
                 patched_data[entry.offset] += base_location;
-        }
+        } // for
 
         data = patched_data;
-    }
+    } // if
 
     const GLuint shader = ctx->glCreateShader(glsl_shader_type(pd->shader_type));
     ctx->glShaderBinary(1, &shader, GL_SHADER_BINARY_FORMAT_SPIR_V_ARB, data, data_len);
@@ -479,7 +479,7 @@
     {
         GLsizei len = 0;
         ctx->glGetShaderInfoLog(shader, sizeof(error_buffer), &len,
-                             (GLchar *)error_buffer);
+                             (GLchar *) error_buffer);
         ctx->glDeleteShader(shader);
         *s = 0;
         return 0;
@@ -492,12 +492,6 @@
 
 static int impl_SPIRV_CompileShader(const MOJOSHADER_parseData *pd, GLuint *s)
 {
-    #if 0
-    FILE *fp = fopen(pd->mainfn, "w");
-    fwrite(pd->output, 1, pd->output_len, fp);
-    fclose(fp);
-    #endif
-
     // Compilation postponed until linking, but generate dummy shader id so hash table lookups work.
     *s = ctx->glCreateShader(glsl_shader_type(pd->shader_type));
     return 1;
@@ -519,14 +513,14 @@
             return 0;
 
         base_location += spv_getPatchTable(vshader)->location_count;
-    }
+    } // if
 
     GLuint ps_handle = 0;
     if (pshader)
     {
         if (!spv_CompileShader(pshader->parseData, base_location, &ps_handle))
             return 0;
-    }
+    } // if
 
     if (ctx->have_opengl_2)
     {
@@ -535,12 +529,12 @@
         {
             ctx->glAttachShader(program, vs_handle);
             ctx->glDeleteShader(vs_handle);
-        }
+        } // if
         if (ps_handle)
         {
             ctx->glAttachShader(program, ps_handle);
             ctx->glDeleteShader(ps_handle);
-        }
+        } // if
         ctx->glLinkProgram(program);
         ctx->glGetProgramiv(program, GL_LINK_STATUS, &ok);
         if (!ok)
@@ -562,12 +556,12 @@
         {
             ctx->glAttachObjectARB(program, (GLhandleARB) vs_handle);
             ctx->glDeleteObjectARB((GLhandleARB) vs_handle);
-        }
+        } // if
         if (ps_handle)
         {
             ctx->glAttachObjectARB(program, (GLhandleARB) ps_handle);
             ctx->glDeleteObjectARB((GLhandleARB) ps_handle);
-        }
+        } // if
         ctx->glLinkProgramARB(program);
         ctx->glGetObjectParameterivARB(program, GL_OBJECT_LINK_STATUS_ARB, &ok);
         if (!ok)
@@ -611,11 +605,9 @@
 {
     GLint base_location = 0;
 
-    const SpirvPatchTable* table = spv_getPatchTable(shader);
+    const SpirvPatchTable *table = spv_getPatchTable(shader);
     if (shader->parseData->shader_type == MOJOSHADER_TYPE_PIXEL)
-    {
         base_location += spv_getPatchTable(program->vertex)->location_count;
-    }
 
     assert(table->samplers[idx].offset);
     return base_location + table->samplers[idx].location;
@@ -623,8 +615,8 @@
 
 static void impl_SPIRV_FinalInitProgram(MOJOSHADER_glProgram *program)
 {
-    const SpirvPatchTable* vs_table = spv_getPatchTable(program->vertex);
-    const SpirvPatchTable* ps_table = spv_getPatchTable(program->fragment);
+    const SpirvPatchTable *vs_table = spv_getPatchTable(program->vertex);
+    const SpirvPatchTable *ps_table = spv_getPatchTable(program->fragment);
     program->vs_float4_loc = vs_table->array_vec4.location;
     program->vs_int4_loc   = vs_table->array_ivec4.location;
     program->vs_bool_loc   = vs_table->array_bool.location;
@@ -641,9 +633,8 @@
         if (program->ps_float4_loc != -1) program->ps_float4_loc += ps_base_location;
         if (program->ps_int4_loc   != -1) program->ps_int4_loc   += ps_base_location;
         if (program->ps_bool_loc   != -1) program->ps_bool_loc   += ps_base_location;
-    }
+    } // if
 } // impl_SPIRV_FinalInitProgram
-
 #endif // SUPPORT_PROFILE_SPIRV
 
 #if SUPPORT_PROFILE_GLSL
@@ -809,7 +800,6 @@
 
 static void impl_GLSL_FinalInitProgram(MOJOSHADER_glProgram *program)
 {
-
     program->vs_float4_loc = glsl_uniform_loc(program, "vs_uniforms_vec4");
     program->vs_int4_loc = glsl_uniform_loc(program, "vs_uniforms_ivec4");
     program->vs_bool_loc = glsl_uniform_loc(program, "vs_uniforms_bool");
diff -r 73770d3b26ae -r 1041c1e58213 utils/testparse.c
--- a/utils/testparse.c	Wed Aug 14 08:07:15 2019 +0200
+++ b/utils/testparse.c	Wed Aug 14 10:29:18 2019 -0400
@@ -46,10 +46,6 @@
 #define Free NULL
 #endif
 
-static FILE* g_out;
-#define printf(...) fprintf(g_out, __VA_ARGS__)
-#define putchar(c) putc((c), g_out)
-
 static inline void do_indent(const unsigned int indent)
 {
     unsigned int i;
@@ -79,15 +75,15 @@
                            unsigned int indent)
 {
     static const char *symclasses[] = {
-            "scalar", "vector", "row-major matrix",
-            "column-major matrix", "object", "struct"
+        "scalar", "vector", "row-major matrix",
+        "column-major matrix", "object", "struct"
     };
 
     static const char *symtypes[] = {
-            "void", "bool", "int", "float", "string", "texture",
-            "texture1d", "texture2d", "texture3d", "texturecube",
-            "sampler", "sampler1d", "sampler2d", "sampler3d",
-            "samplercube", "pixelshader", "vertexshader", "unsupported"
+        "void", "bool", "int", "float", "string", "texture",
+        "texture1d", "texture2d", "texture3d", "texturecube",
+        "sampler", "sampler1d", "sampler2d", "sampler3d",
+        "samplercube", "pixelshader", "vertexshader", "unsupported"
     };
 
     INDENT();
@@ -132,7 +128,7 @@
         for (i = 0; i < symbol_count; i++, sym++)
         {
             static const char *regsets[] = {
-                    "bool", "int4", "float4", "sampler"
+                "bool", "int4", "float4", "sampler"
             };
 
             INDENT(); printf("    * %d: \"%s\"\n", i, sym->name);
@@ -223,10 +219,10 @@
     int i, j;
 
     static const char *opcodestr[] = {
-            "nop", "mov", "neg", "rcp", "frc", "exp", "log", "rsq", "sin", "cos",
-            "asin", "acos", "atan", "min", "max", "lt", "ge", "add", "mul",
-            "atan2", "div", "cmp", "movc", "dot", "noise", "min", "max", "lt",
-            "ge", "add", "mul", "atan2", "div", "dot", "noise"
+        "nop", "mov", "neg", "rcp", "frc", "exp", "log", "rsq", "sin", "cos",
+        "asin", "acos", "atan", "min", "max", "lt", "ge", "add", "mul",
+        "atan2", "div", "cmp", "movc", "dot", "noise", "min", "max", "lt",
+        "ge", "add", "mul", "atan2", "div", "dot", "noise"
     };
 
     INDENT(); printf("PRESHADER:\n");
@@ -267,10 +263,11 @@
         printf("\n");
         for (i = 0; i < count; i++)
         {
-            static const char *usagenames[] = { "<unknown>",
-                    "position", "blendweight", "blendindices", "normal",
-                    "psize", "texcoord", "tangent", "binormal", "tessfactor",
-                    "positiont", "color", "fog", "depth", "sample"
+            static const char *usagenames[] = {
+                "<unknown>",
+                "position", "blendweight", "blendindices", "normal",
+                "psize", "texcoord", "tangent", "binormal", "tessfactor",
+                "positiont", "color", "fog", "depth", "sample"
             };
             const MOJOSHADER_attribute *a = &attributes[i];
             char numstr[16] = { 0 };
@@ -287,7 +284,7 @@
 
 
 static void print_shader(const char *fname, const MOJOSHADER_parseData *pd,
-                         unsigned int indent, const char* srcfile)
+                         unsigned int indent)
 {
     INDENT(); printf("PROFILE: %s\n", pd->profile);
     if (pd->error_count > 0)
@@ -327,12 +324,12 @@
                 if (c->type == MOJOSHADER_UNIFORM_FLOAT)
                 {
                     printf("%f %f %f %f", c->value.f[0], c->value.f[1],
-                           c->value.f[2], c->value.f[3]);
+                                          c->value.f[2], c->value.f[3]);
                 } // if
                 else if (c->type == MOJOSHADER_UNIFORM_INT)
                 {
                     printf("%d %d %d %d", c->value.i[0], c->value.i[1],
-                           c->value.i[2], c->value.i[3]);
+                                          c->value.i[2], c->value.i[3]);
                 } // else if
                 else if (c->type == MOJOSHADER_UNIFORM_BOOL)
                 {
@@ -406,36 +403,14 @@
         {
             const char *output;
             int output_len;
-            int i, is_spirv;
+            int i;
 
-            is_spirv = strcmp(pd->profile, "spirv") == 0;
-
-            if (is_spirv)
+            if (strcmp(pd->profile, "spirv") == 0)
             {
 #if SUPPORT_PROFILE_SPIRV && defined(MOJOSHADER_HAS_SPIRV_TOOLS)
                 int binary_len = pd->output_len - sizeof(SpirvPatchTable);
-                #if 0
-                const char *shader_type = NULL;
-                switch (pd->shader_type)
-                {
-                    case MOJOSHADER_TYPE_PIXEL: shader_type = "ps"; break;
-                    case MOJOSHADER_TYPE_VERTEX: shader_type = "vs"; break;
-                    case MOJOSHADER_TYPE_GEOMETRY: shader_type = "gs"; break;
-                    default: assert(!"Invalid shader type");
-                }
-                #endif
 
-                #if 0
-                {
-                    char filename[512];
-                    snprintf(filename, sizeof(filename), "%s.%s.%s.spv", srcfile, pd->mainfn, shader_type);
-                    FILE* f = fopen(filename, "wb");
-                    fwrite(pd->output, 1, binary_len, f);
-                    fclose(f);
-                }
-                #endif
-
-                uint32_t *words = (uint32_t *)pd->output;
+                uint32_t *words = (uint32_t *) pd->output;
                 size_t word_count = binary_len / 4;
 
                 spv_text text;
@@ -447,16 +422,6 @@
                 {
                     output = text->str;
                     output_len = text->length;
-
-                    #if 0
-                    {
-                        char filename[512];
-                        snprintf(filename, sizeof(filename), "%s.%s.spvasm", pd->mainfn, shader_type);
-                        FILE* fdis = fopen(filename, "wt");
-                        fwrite(output, 1, output_len, fdis);
-                        fclose(fdis);
-                    }
-                    #endif
                 } // if
                 else
                 {
@@ -469,16 +434,6 @@
                     fprintf(stderr, "\nVALIDATION FAILURE: %s\n\n", diagnostic->error);
                 } // if
 
-                #if 0
-                {
-                    char filename[512];
-                    snprintf(filename, sizeof(filename), "%s.%s.spvasm", pd->mainfn, shader_type);
-                    FILE* fdis = fopen(filename, "wb");
-                    fwrite(pd->output, 1, binary_len, fdis);
-                    fclose(fdis);
-                }
-                #endif
-
                 if (disResult != SPV_SUCCESS || validateResult != SPV_SUCCESS)
                 {
                     exit(EXIT_FAILURE);
@@ -534,35 +489,35 @@
     printf("VALUE: %s -> %s\n", value->name, value->semantic);
 
     static const char *classes[] =
-            {
-                    "SCALAR",
-                    "VECTOR",
-                    "ROW-MAJOR MATRIX",
-                    "COLUMN-MAJOR MATRIX",
-                    "OBJECT",
-                    "STRUCT"
-            };
+    {
+        "SCALAR",
+        "VECTOR",
+        "ROW-MAJOR MATRIX",
+        "COLUMN-MAJOR MATRIX",
+        "OBJECT",
+        "STRUCT"
+    };
     static const char *types[] =
-            {
-                    "VOID",
-                    "BOOL",
-                    "INT",
-                    "FLOAT",
-                    "STRING",
-                    "TEXTURE",
-                    "TEXTURE1D",
-                    "TEXTURE2D",
-                    "TEXTURE3D",
-                    "TEXTURECUBE",
-                    "SAMPLER",
-                    "SAMPLER1D",
-                    "SAMPLER2D",
-                    "SAMPLER3D",
-                    "SAMPLERCUBE",
-                    "PIXELSHADER",
-                    "VERTEXSHADER",
-                    "UNSUPPORTED"
-            };
+    {
+        "VOID",
+        "BOOL",
+        "INT",
+        "FLOAT",
+        "STRING",
+        "TEXTURE",
+        "TEXTURE1D",
+        "TEXTURE2D",
+        "TEXTURE3D",
+        "TEXTURECUBE",
+        "SAMPLER",
+        "SAMPLER1D",
+        "SAMPLER2D",
+        "SAMPLER3D",
+        "SAMPLERCUBE",
+        "PIXELSHADER",
+        "VERTEXSHADER",
+        "UNSUPPORTED"
+    };
     do_indent(indent + 1);
     printf("CLASS: %s\n", classes[value->type.parameter_class]);
     do_indent(indent + 1);
@@ -575,10 +530,10 @@
     printf("TOTAL VALUES: %d\n", value->value_count);
 
     if (value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER
-        || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER1D
-        || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER2D
-        || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER3D
-        || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLERCUBE)
+     || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER1D
+     || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER2D
+     || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLER3D
+     || value->type.parameter_type == MOJOSHADER_SYMTYPE_SAMPLERCUBE)
     {
         do_indent(indent + 1);
         printf("SAMPLER VALUES:\n");
@@ -587,26 +542,26 @@
             MOJOSHADER_effectSamplerState *state = &value->valuesSS[i];
 
             static const char *samplerstatetypes[] =
-                    {
-                            "UNKNOWN0",
-                            "UNKNOWN1",
-                            "UNKNOWN2",
-                            "UNKNOWN3",
-                            "TEXTURE",
-                            "ADDRESSU",
-                            "ADDRESSV",
-                            "ADDRESSW",
-                            "BORDERCOLOR",
-                            "MAGFILTER",
-                            "MINFILTER",
-                            "MIPFILTER",
-                            "MIPMAPLODBIAS",
-                            "MAXMIPLEVEL",
-                            "MAXANISOTROPY",
-                            "SRGBTEXTURE",
-                            "ELEMENTINDEX",
-                            "DMAPOFFSET",
-                    };
+            {
+                "UNKNOWN0",
+                "UNKNOWN1",
+                "UNKNOWN2",
+                "UNKNOWN3",
+                "TEXTURE",
+                "ADDRESSU",
+                "ADDRESSV",
+                "ADDRESSW",
+                "BORDERCOLOR",
+                "MAGFILTER",
+                "MINFILTER",
+                "MIPFILTER",
+                "MIPMAPLODBIAS",
+                "MAXMIPLEVEL",
+                "MAXANISOTROPY",
+                "SRGBTEXTURE",
+                "ELEMENTINDEX",
+                "DMAPOFFSET",
+            };
             do_indent(indent + 2);
             printf("TYPE: %s -> ", samplerstatetypes[state->type]);
 
@@ -631,26 +586,26 @@
         do
         {
             static const char *prints[] =
-                    {
-                            "%X ",
-                            "%d ",
-                            "%d ",
-                            "%.2f ",
-                            "%d ",
-                            "%d ",
-                            "%d ",
-                            "%d ",
-                            "%d ",
-                            "%d ",
-                            "SAMPLER?! ",
-                            "SAMPLER?! ",
-                            "SAMPLER?! ",
-                            "SAMPLER?! ",
-                            "SAMPLER?! ",
-                            "%d ",
-                            "%d ",
-                            "%X "
-                    };
+            {
+                "%X ",
+                "%d ",
+                "%d ",
+                "%.2f ",
+                "%d ",
+                "%d ",
+                "%d ",
+                "%d ",
+                "%d ",
+                "%d ",
+                "SAMPLER?! ",
+                "SAMPLER?! ",
+                "SAMPLER?! ",
+                "SAMPLER?! ",
+                "SAMPLER?! ",
+                "%d ",
+                "%d ",
+                "%X "
+            };
             for (r = 0; r < value->type.rows; r++)
             {
                 do_indent(indent + 2);
@@ -670,7 +625,7 @@
 
 
 static void print_effect(const char *fname, const MOJOSHADER_effect *effect,
-                         const unsigned int indent, const char* srcfile)
+                         const unsigned int indent)
 {
     INDENT();
     printf("PROFILE: %s\n", effect->profile);
@@ -683,8 +638,8 @@
             const MOJOSHADER_error *err = &effect->errors[i];
             INDENT();
             printf("%s:%d: ERROR: %s\n",
-                   err->filename ? err->filename : fname,
-                   err->error_position, err->error);
+                    err->filename ? err->filename : fname,
+                    err->error_position, err->error);
         } // for
     } // if
     else
@@ -738,7 +693,7 @@
         {
             INDENT();
             if (object->type == MOJOSHADER_SYMTYPE_PIXELSHADER
-                || object->type == MOJOSHADER_SYMTYPE_VERTEXSHADER)
+             || object->type == MOJOSHADER_SYMTYPE_VERTEXSHADER)
             {
                 if (object->shader.is_preshader)
                 {
@@ -751,24 +706,24 @@
                 {
                     printf("OBJECT #%d: SHADER, technique %u, pass %u\n", i,
                            object->shader.technique, object->shader.pass);
-                    print_shader(fname, object->shader.shader, indent + 1, srcfile);
+                    print_shader(fname, object->shader.shader, indent + 1);
                 } // else
             } // if
             else if (object->type == MOJOSHADER_SYMTYPE_STRING)
                 printf("OBJECT #%d: STRING, '%s'\n", i,
                        object->string.string);
             else if (object->type == MOJOSHADER_SYMTYPE_SAMPLER
-                     || object->type == MOJOSHADER_SYMTYPE_SAMPLER1D
-                     || object->type == MOJOSHADER_SYMTYPE_SAMPLER2D
-                     || object->type == MOJOSHADER_SYMTYPE_SAMPLER3D
-                     || object->type == MOJOSHADER_SYMTYPE_SAMPLERCUBE)
+                  || object->type == MOJOSHADER_SYMTYPE_SAMPLER1D
+                  || object->type == MOJOSHADER_SYMTYPE_SAMPLER2D
+                  || object->type == MOJOSHADER_SYMTYPE_SAMPLER3D
+                  || object->type == MOJOSHADER_SYMTYPE_SAMPLERCUBE)
                 printf("OBJECT #%d: MAPPING, '%s'\n", i,
                        object->mapping.name);
             else if (object->type == MOJOSHADER_SYMTYPE_TEXTURE
-                     || object->type == MOJOSHADER_SYMTYPE_TEXTURE1D
-                     || object->type == MOJOSHADER_SYMTYPE_TEXTURE2D
-                     || object->type == MOJOSHADER_SYMTYPE_TEXTURE3D
-                     || object->type == MOJOSHADER_SYMTYPE_TEXTURECUBE)
+                  || object->type == MOJOSHADER_SYMTYPE_TEXTURE1D
+                  || object->type == MOJOSHADER_SYMTYPE_TEXTURE2D
+                  || object->type == MOJOSHADER_SYMTYPE_TEXTURE3D
+                  || object->type == MOJOSHADER_SYMTYPE_TEXTURECUBE)
                 printf("OBJECT #%d: TEXTURE\n", i);
             else
                 printf("UNKNOWN OBJECT: #%d\n", i);
@@ -781,7 +736,7 @@
 
 
 static int do_parse(const char *fname, const unsigned char *buf,
-                    const int len, const char *prof, const char* srcfile)
+                    const int len, const char *prof)
 {
     int retval = 0;
 
@@ -808,13 +763,15 @@
                         const MOJOSHADER_parseData *shader = object->shader.shader;
                         if (shader)
                             error_count += shader->error_count;
-                    }
+                    } // if
+                    break;
+                default:
                     break;
             }
         }
         retval = (error_count == 0);
         printf("EFFECT:\n");
-        print_effect(fname, effect, 1, srcfile);
+        print_effect(fname, effect, 1);
         MOJOSHADER_freeEffect(effect);
 #else
         printf("Is an effect, but effect support is disabled!\n");
@@ -828,66 +785,49 @@
                               NULL, 0, Malloc, Free, NULL);
         retval = (pd->error_count == 0);
         printf("SHADER:\n");
-        print_shader(fname, pd, 1, srcfile);
+        print_shader(fname, pd, 1);
         MOJOSHADER_freeParseData(pd);
     } // else
 
     return retval;
 } // do_parse
 
-static void print_logo(FILE* f)
+static void print_logo()
 {
-    fprintf(f, "MojoShader testparse\n");
-    fprintf(f, "Compiled against changeset %s\n", MOJOSHADER_CHANGESET);
-    fprintf(f, "Linked against changeset %s\n", MOJOSHADER_changeset());
-    fprintf(f, "\n");
+    printf("MojoShader testparse\n");
+    printf("Compiled against changeset %s\n", MOJOSHADER_CHANGESET);
+    printf("Linked against changeset %s\n", MOJOSHADER_changeset());
+    printf("\n");
 }
 
 int main(int argc, char **argv)
 {
     int retval = 0;
 
-    g_out = stdout;
-
     int no_logo = 0;
     int i = 1;
     while (argc - i >= 2)
     {
-        if (0 == strcmp(argv[i], "-o"))
-        {
-            FILE* f = fopen(argv[i + 1], "wb");
-            if (f)
-            {
-                if (g_out != stdout)
-                    fclose(g_out);
-
-                g_out = f;
-            }
-
-            i += 2;
-        }
-        else if (0 == strcmp(argv[i], "--no-logo"))
+        if (strcmp(argv[i], "--no-logo") == 0)
         {
             no_logo = 1;
             i += 1;
-        }
+        } // if
         else
-        {
             break;
-        }
     }
 
-    if (argc - i < 2)
+    if ((argc - i) < 2)
     {
         if (!no_logo)
-            print_logo(stdout);
+            print_logo();
 
         fprintf(stdout, "\n\nUSAGE: %s [-o <dstFile>] <profile> [file1] ... [fileN]\n\n", argv[0]);
-    }
+    } // if
     else
     {
         if (!no_logo)
-            print_logo(g_out);
+            print_logo();
 
         const char *profile = argv[i++];
         for (; i < argc; i++)
@@ -897,24 +837,19 @@
             {
                 fprintf(stderr, "%s\n", argv[i]);
                 fprintf(stderr, " ... fopen('%s') failed.\n", argv[i]);
-            }
+            } // if
             else
             {
                 unsigned char *buf = (unsigned char *) malloc(1000000);
                 int rc = fread(buf, 1, 1000000, io);
                 fclose(io);
-                if (!do_parse(argv[i], buf, rc, profile, argv[i]))
+                if (!do_parse(argv[i], buf, rc, profile))
                     retval = 1;
                 free(buf);
             } // else
         } // for
     } // else
 
-    if (g_out != stdout)
-    {
-        fclose(g_out);
-    }
-
     return retval;
 } // main
 
```

**Comment by krolli** [comment]
Patch applied.

> Probably have to implement the vpos flip variable (no need to adjust half-pixel!)

Was planning to take a look at it after I saw the change on default branch.

> Can we remove the ARB program paths from SPIR-V? Seems unlikely that a driver will support ARB_gl_spirv but not GL2...

Yeah, I don't think we'll run into a driver like that either. It is just a couple lines so I didn't think too much about it (and I kept it for parity with glsl), but it can be removed.

I was planning to remove most of the debug stuff myself, but I guess you beat me to it. On git I kept debug things on separate branch, but it was pain to work with and I figured it would be easier to remove it later.

I will miss having testparse spit output deterministically into a file using command line args instead of redirecting stdout+stderr, but I can always use some of the older revisions for that. :)

**Comment by flibitijibibo** [comment]
Part 2 is here! This is 100% pure style cleanup, no functionality or structural changes were made. I only had one note for this pass: We should move all the internal `static spv_` functions to the top, maybe inline them as well. But, for now we're going to run the testing with this stuff exactly as-is, just in case it makes life easier before we start moving functions around.

http://www.flibitijibibo.com/spvcp2.diff

Expect the announcement to show up on my plan file...

https://icculus.org/finger/flibitijibibo

**Comment by flibitijibibo** [comment]
Quick update with a patch to fix the CentOS build...

http://www.flibitijibibo.com/spvcp3.diff

**Comment by flibitijibibo** [comment]
Spamming this thread a lot today... sorry about that >_>

I just checked my test list and it looks like we're almost good to go; I need to update FEZ to 1.13 but other than that...

```
FEZ		GOOD, NEEDS 1.13
Dust: AET	GOOD
Murder Miners	NOTGOOD
Escape Goat 2	NOTGOOD
Capsized	GOOD
Apotheon	GOOD
Cryptark	GOOD
Bastion		GOOD
Reus		GOOD
Celeste		GOOD
```

... all you have to do is extract my [custom FNA build](http://www.flibitijibibo.com/files/fnaSPIRV.tar.bz2) to the game directory and it'll work. You can even set `FNATEST_DISABLE_SPIRV=1` to be sure GLSL still works.

The only thing hanging us up now is `vpFlip` not working, seems NVIDIA Linux generates GL_INVALID_OPERATION in glProgramViewportInfo. Once that's fixed and all the patches are in I'll do a new build and get this into testing!

**Comment by flibitijibibo** [comment]
We hacked on this in the Discord and now the builds work! Expect the plan file update tomorrow (maybe).

**Comment by flibitijibibo** [comment]
Just pushed some updates to upstream that affects the GLSL texld instructions. texld is still kinda eh...

https://github.com/FNA-XNA/MojoShader/blob/fna/profiles/mojoshader_profile_glsl.c#L1696-L1819

... but check out texldd/texldl:

https://github.com/FNA-XNA/MojoShader/blob/fna/profiles/mojoshader_profile_glsl.c#L2246-L2300

https://github.com/FNA-XNA/MojoShader/blob/fna/profiles/mojoshader_profile_glsl.c#L2326-L2373

If anyone wants those functions, hopefully this is WAY easier to sit through.

**Comment by flibitijibibo** [comment]
Sent krolli some patches for TEXLD/TEXLDL/TEXLDD. They seemed simple enough, so I think those are okay...? (Unless I screwed up the source arg loading, which it sounds like I did.)

Something really amusing that happened this week was D9VK's 0.20 release having [the ~exact~ list of instructions that we ignored](https://github.com/Joshua-Ashton/d9vk/releases/tag/0.20) for this bounty, may be worth exploring:

- ~TEXBEM~ Wait, no, we have that, just TEXBEML is missing...
- [TEXM3X3TEX](https://github.com/Joshua-Ashton/d9vk/commit/06f06700262309e80560c48b9036170a8630dd65)

**Comment by flibitijibibo** [comment]
TEXM3X3SPEC got done in the latest d9vk.

**Comment by krolli** [comment]
Went over the remaining opcodes and code paths I knew were missing for parity with glsl emittor and got them out of the way. In other words:
- added missing source modifiers `BIASNEGATE`, `SIGNNEGATE`, `X2`, `X2NEGATE`
- added missing opcodes `TEXBEML`, `TEXM3X3`, `TEXM3X3TEX`, `TEXM3X3SPEC`

I also have some minimal shaders that use those instructions and I think there might be some bugs in GLSL emitter and even mojoshader parser when dealing with some of those. I will have to take a better look at things to be sure though.

**Comment by flibitijibibo** [comment]
SPIR-V support has been merged! Thanks to everyone who helped out with this.
